fud2/src/main.rs

```diff
@@ -145,21 +149,97 @@ fn build_driver() -> Driver {
+            e.arg("backend", "firrtl")?;
+            Ok(())
+        },
+    );
+    let firrtl_verilog_setup = bld.setup("Firrtl to Verilog compiler", |e| {
+        e.config_var("firrtl_exe", "firrtl.exe")?;
+        e.rule("firrtl", "$firrtl_exe -i $in -o $out -X sverilog")?;
+        Ok(())
+    });
+
+    // Helper function for the Calyx --> FIRRTL --> System-Verilog pipeline
+    fn emit_verilog_via_firrtl(e: &mut Emitter, input: &str, output: &str) -> EmitResult {
+        // Generate the FIRRTL
+        let firrtl_name = "sim.fir";
+        e.build_cmd(&[firrtl_name], "calyx", &[input], &[])?;
+        e.arg("backend", "firrtl")?;
+
+        // Compile the FIRRTL into Verilog
+        e.build_cmd(&[output], "firrtl", &[firrtl_name], &[])?;
+
+        Ok(())
+    }
+
+    bld.op(
+        "firrtl",
+        &[calyx_setup, firrtl_verilog_setup],
+        calyx,
+        verilog,
+        emit_verilog_via_firrtl,
+    );
+
+    // Run the whole Calyx --> FIRRTL --> System-Verilog --> Execution via Icarus-Verilog pipeline
+    bld.op(
+        "icarus-firrtl",
+        &[calyx_setup, firrtl_verilog_setup, sim_setup, icarus_setup],
+        calyx,
+        dat,
+        |e, input, output| {
+            let verilog_name = "sim.sv";
+            emit_verilog_via_firrtl(e, input, verilog_name)?;
+
+            // borrowed the below from emit_icarus
+            // Compile the Verilog.
+            let bin_name = "icarus_bin";
+            e.build("icarus-compile", verilog_name, bin_name)?;
+            e.arg(
+                "extra_primitives",
+                &format!("{}/primitives-for-firrtl.sv", e.config_val("data")?),
+            )?;
+            emit_sim_run(e, bin_name, output, false)
+        },
+    );
+
     // Verilator.
     let verilator_setup = bld.setup("Verilator", |e| {
         e.config_var_or("verilator", "verilator.exe", "verilator")?;
         e.config_var_or("cycle_limit", "sim.cycle_limit", "500000000")?;
         e.rule(
             "verilator-compile",
-            "$verilator $in $testbench --trace --binary --top-module TOP -fno-inline -Mdir $out_dir",
+            "$verilator $in $testbench $extra_primitives --trace --binary --top-module TOP -fno-inline -Mdir $out_dir",
         )?;
         Ok(())
     });
-    fn emit_verilator(e: &mut Emitter, input: &str, output: &str, trace: bool) -> EmitResult {
+    fn emit_verilator(
+        e: &mut Emitter,
+        input: &str,
+        output: &str,
+        trace: bool,
+        firrtl: bool,
+    ) -> EmitResult {
         let out_dir = "verilator-out";
         let sim_bin = format!("{}/VTOP", out_dir);
         e.build("verilator-compile", input, &sim_bin)?;
         e.arg("out_dir", out_dir)?;
+        if firrtl {
+            e.arg(
+                "extra_primitives",
+                &format!("{}/primitives-for-firrtl.sv", e.config_val("data")?),
+            )?;
+        } else {
+            e.arg("extra_primitives", "")?;
```

Comment:
FWIW, you can probably leave off these "empty" assignments; in Ninja, variables that are unassigned are just empty by default.

Reply:
Ah I see! Thanks for letting me know :)